src/lyrics/lyrics.cc

```diff
@@ -0,0 +1,375 @@
+
+void update_lyrics_window (const char * title, const char * artist, const char * lyrics)
+{
+    GtkTextIter iter;
+
+    if (! textbuffer)
+        return;
+
+    gtk_text_buffer_set_text (textbuffer, "", -1);
+
+    gtk_text_buffer_get_start_iter (textbuffer, & iter);
+
+    gtk_text_buffer_insert_with_tags_by_name (textbuffer, & iter, title, -1,
+     "weight_bold", "scale_large", nullptr);
+
+    if (artist)
+    {
+        gtk_text_buffer_insert (textbuffer, & iter, "\n", -1);
+        gtk_text_buffer_insert_with_tags_by_name (textbuffer, & iter, artist, -1,
+         "style_italic", nullptr);
+    }
+
+    gtk_text_buffer_insert (textbuffer, & iter, "\n\n", -1);
+    gtk_text_buffer_insert (textbuffer, & iter, lyrics, -1);
+
+    gtk_text_buffer_get_start_iter (textbuffer, & iter);
+    gtk_text_view_scroll_to_iter (textview, & iter, 0, true, 0, 0);
+}
+
+bool try_parse_json (const Index<char> & buf, const char * key, String & output)
+{
+    JsonParser * parser = json_parser_new ();
+
+    if (! json_parser_load_from_data (parser, buf.begin (), buf.len (), nullptr))
+    {
+        g_object_unref (parser);
+        return false;
+    }
+
+    JsonNode * root = json_parser_get_root (parser);
+    JsonReader * reader = json_reader_new (root);
+
+    json_reader_read_member (reader, key);
+    output = String (json_reader_get_string_value (reader));
+    json_reader_end_member (reader);
+
+    g_object_unref (reader);
+    g_object_unref (parser);
+
+    return true;
+}
+
+static char * truncate_by_pattern (const char * input, const char * pattern)
+{
+    GRegex * regex = g_regex_new (pattern, G_REGEX_CASELESS, (GRegexMatchFlags) 0, nullptr);
+    char * result = g_regex_replace (regex, input, strlen (input), 0, "", (GRegexMatchFlags) 0, nullptr);
+    g_regex_unref (regex);
+    return result;
+}
+
+static void split_title_and_truncate ()
+{
+    StringBuf split_pattern = str_concat ({
+        "^(.*)\\s+[", aud_get_str (CFG_SECTION, "split-on-chars"), "]\\s+(.*)$"
+    });
+
+    GMatchInfo * match_info;
+    GRegex * split_regex = g_regex_new (split_pattern, G_REGEX_CASELESS, (GRegexMatchFlags) 0, nullptr);
+
+    if (g_regex_match (split_regex, g_state.title, (GRegexMatchFlags) 0, & match_info))
+    {
+        CharPtr artist (g_match_info_fetch (match_info, 1));
+        CharPtr title (g_match_info_fetch (match_info, 2));
+
+        if (aud_get_bool (CFG_SECTION, "truncate-fields-on-chars"))
+        {
+            StringBuf artist_pattern = str_concat ({
+                "^.*\\s+[", aud_get_str (CFG_SECTION, "truncate-on-chars"), "]\\s+"
+            });
+
+            StringBuf title_pattern = str_concat ({
+               "\\s+[", aud_get_str (CFG_SECTION, "truncate-on-chars"), "]\\s+.*$"
+            });
+
+            // TODO: Are these reassignments safe and not leaking memory?
+            artist = CharPtr (truncate_by_pattern (artist, artist_pattern));
+            title = CharPtr (truncate_by_pattern (title, title_pattern));
```

Comment:
I assume the results of `g_match_info_fetch()` are leaked here. Is this correct? If yes, what's the cleanest way to free and reassign the strings?

Reply:
This is actually fine, nothing is leaked. `CharPtr` has overloaded `operator=` which handles freeing the previous pointer.